siva/storage.go

```diff
@@ -0,0 +1,170 @@
+package siva
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	sivafs "gopkg.in/src-d/go-billy-siva.v4"
+	billy "gopkg.in/src-d/go-billy.v4"
+	butil "gopkg.in/src-d/go-billy.v4/util"
+	"gopkg.in/src-d/go-git.v4/plumbing/cache"
+	"gopkg.in/src-d/go-git.v4/storage"
+	"gopkg.in/src-d/go-git.v4/storage/filesystem"
+	"gopkg.in/src-d/go-git.v4/storage/transactional"
+)
+
+const (
+	transactionName = "transaction"
+	baseSivaName    = "base"
+)
+
+// Storage holds a ReadWrite siva storage. It can be transactional in which
+// case it will write to a temporary siva file and will append it to the
+// original siva on Commit.
+type Storage struct {
+	storage.Storer
+
+	base   billy.Filesystem
+	path   string
+	fs     sivafs.SivaFS
+	tmp    billy.Filesystem
+	tmpDir string
+}
+
+// NewStorage creates a new Storage struct. A new temporary directory is created
+// for the siva filesystem that can be later deleted with Cleanup.
+func NewStorage(
+	base billy.Filesystem,
+	path string,
```

Comment:
what for you need a path?

Reply:
It needs to know where the siva file is to open and update it.